(FACT-2246) Create mountpoints resolver

GabrielNagy opened the pull request:
Added dependency to `sys-filesystem` which uses FFI to get filesystem information. Also fixed a bug where `BytesToHumanReadable` did not add 2 digits after the decimal point.

GabrielNagy committed:
(FACT-2246) Create mountpoints resolver

GabrielNagy commented:
```sh-session
> ./bin/facter-ng mountpoints > mountpoints.facter-ng
> /usr/bin/facter mountpoints > mountpoints.facter
> diff -u mountpoints.facter mountpoints.facter-ng
--- mountpoints.facter	2019-12-17 15:51:52.406898271 +0200
+++ mountpoints.facter-ng	2019-12-17 15:51:49.073564901 +0200
@@ -1,8 +1,8 @@
 {
   / => {
-    available => "86.80 GiB",
-    available_bytes => 93201674240,
-    capacity => "80.02%",
+    available => "64.67 GiB",
+    available_bytes => 69435973632,
+    capacity => "84.31%",
     device => "/dev/nvme0n1p2",
     filesystem => "ext4",
     options => [
```

The only difference seems to be the root-reserved thing from https://github.com/puppetlabs/facter/pull/1857
[mountpoints.facter.txt](https://github.com/puppetlabs/facter-ng/files/3973580/mountpoints.facter.txt)
[mountpoints.facter-ng.txt](https://github.com/puppetlabs/facter-ng/files/3973581/mountpoints.facter-ng.txt)

GabrielNagy commented on .rubocop.yml:
```diff
@@ -23,6 +23,7 @@ Naming/ClassAndModuleCamelCase:
     - 'spec/mocks/**/*'
 
 Metrics/AbcSize:
+  Max: 30
```
i would disable this cop entirely since it's really subjective based on the case

sebastian-miclea commented on lib/resolvers/mountpoints_resolver.rb:
```diff
@@ -0,0 +1,77 @@
+              device = fs.name
+              filesystem = fs.mount_type
+              path = fs.mount_point
+              options = fs.options.split(',')
```
Are the options used anywhere?

GabrielNagy replied:
Yeah, they get put in the hash at line 69. They're part of `MOUNT_KEYS`.

sebastian-miclea commented on lib/resolvers/mountpoints_resolver.rb:
```diff
@@ -0,0 +1,77 @@
+
+              # If the "root" device, lookup the actual device from the kernel options
+              # This is done because not all systems symlink /dev/root
+              device = root_device if device == '/dev/root'
```
The code would look a bit cleaner if the root_device method receives **device** as an argument, and decide inside the method what the device is.

GabrielNagy replied:
The thing is, we want to call the `root_device` method only if our device is `/dev/root`. We can't have any other valid argument. We also don't care about `device` itself in the method.

sebastian-miclea replied:
Might not phrased properly, but I would see this as a method that handles all devices, not only root. Basically, move the if inside and call it always. 

`device = compute_device(fs.name)`
This would reduce complexity of the **read_mounts** method

sebastian-miclea commented on lib/resolvers/mountpoints_resolver.rb:
```diff
@@ -0,0 +1,77 @@
+              device = root_device if device == '/dev/root'
+
+              stats = Sys::Filesystem.stat(path)
+              size_bytes = stats.bytes_total
```
You could send directly stats.bytes_total to the convert method without assigning to a local var.

GabrielNagy replied:
Makes sense, but `size_bytes` is also part of the final fact, so I need it in its own variable.

sebastian-miclea commented on lib/resolvers/mountpoints_resolver.rb:
```diff
@@ -0,0 +1,77 @@
+              options = fs.options.split(',')
+
+              next if path =~ %r{^/(proc|sys)} && filesystem != 'tmpfs'
+              next if filesystem == 'autofs'
```
I'm not sure what type is **fs** is but, couldn't we reject these before iterating?

GabrielNagy replied:
I don't think so, we always get all mounts from `Sys::Filesystem.mounts`. We can drop unwanted mounts but then we'd have to iterate twice; first to reject and then to parse.

GabrielNagy committed:
(FACT-2246) Add tests for mountpoints resolver

Mock `sys/filesystem` due to the use of FFI. Fix redundant escape in
`root_device` regex. Move the `root_device` check in a separate,
extensible `compute_device` method.

GabrielNagy committed:
(FACT-2246) Revert .rubocop.yml

GabrielNagy committed:
(FACT-2246) Shorten read_mounts method

GabrielNagy commented:
I pushed some tests and reverted the `.rubocop.yml` changes. There's only the `Metrics/AbcSize` cop now.

```
Inspecting 769 files
.......................................................................................................................................................................................................................................................................................................................................................................C.........................................................................................................................................................................................................................................................................................................................................................................................................................

Offenses:

lib/resolvers/mountpoints_resolver.rb:48:11: C: Metrics/AbcSize: Assignment Branch Condition size for read_mounts is too high. [28.74/15]
          def read_mounts ...
          ^^^^^^^^^^^^^^^

769 files inspected, 1 offense detected
```

GabrielNagy commented:
Also specs are not running anymore after rebasing:

```
An error occurred while loading spec_helper.
Failure/Error:
  def initialize(logged_class)
    determine_callers_name(logged_class)
    set_format_for_file_logger
    set_format_for_stdout
  end

ArgumentError:
  wrong number of arguments (given 0, expected 1)
# ./lib/framework/logging/logger.rb:10:in `initialize'
# ./lib/resolvers/mountpoints_resolver.rb:9:in `new'
# ./lib/resolvers/mountpoints_resolver.rb:9:in `<class:Mountpoints>'
# ./lib/resolvers/mountpoints_resolver.rb:6:in `<module:Linux>'
# ./lib/resolvers/mountpoints_resolver.rb:5:in `<module:Resolvers>'
# ./lib/resolvers/mountpoints_resolver.rb:4:in `<module:Facter>'
# ./lib/resolvers/mountpoints_resolver.rb:3:in `<top (required)>'
# ./spec/mocks/kernel_mock.rb:7:in `require'
# ./spec/mocks/kernel_mock.rb:7:in `require'
# ./lib/framework/core/file_loader.rb:15:in `glob'
# ./lib/framework/core/file_loader.rb:15:in `load_dir'
# ./lib/framework/core/file_loader.rb:19:in `load_lib_dirs'
# ./lib/framework/core/file_loader.rb:39:in `<top (required)>'
# ./spec/mocks/kernel_mock.rb:7:in `require'
# ./spec/mocks/kernel_mock.rb:7:in `require'
# ./lib/facter-ng.rb:7:in `<top (required)>'
# ./spec/mocks/kernel_mock.rb:7:in `require'
# ./spec/mocks/kernel_mock.rb:7:in `require'
# ./spec/spec_helper.rb:21:in `<top (required)>'
No examples found.
No examples found.


Finished in 0.00003 seconds (files took 0.37146 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples
```

GabrielNagy committed:
(FACT-2246) Spec for Fedora mountpoints fact

GabrielNagy committed:
(FACT-2246) Remove pry-byebug from gemspec

JRuby does not support byebug.

GabrielNagy committed:
(FACT-2246) Fix `compute_capacity`
